### PR TITLE
Require black and isort for contributions

### PR DESCRIPTION
**Pull request by srittau**
* Add explanation to CONTRIBUTNG.md
* Add sample pre-commit script
* Check for correctly formatted files in CI
* Only install flake8 packages when running flake8 in CI

This is a draft, which we should merge only after the following are solved:

* There is a problem with isort that causes it to crash on stdlib/3/io.pyi. Reported as timothycrosley/isort#1027.
* We need to reformat all files at once. My suggestion is to make sure that the PR to do that is open only for a short time, ideally only as long as CI needs, to minimize the disruption.

### CODE REVIEW

File: stdlib/3/encodings/__init__.pyi [hunk 1]
```diff
@@ -1,3 +1,6 @@
+# This commit serves as a workaround for timothycrosley/isort#1027 and can
```
Comment (JelleZijlstra) [inline stdlib/3/encodings/__init__.pyi]: commit -> comment?

File: .travis.yml [hunk 2]
```diff
@@ -21,5 +21,11 @@ jobs:
     - name: "check file consistency"
       script: ./tests/check_consistent.py
     - name: "flake8"
-      install: pip install -r requirements-tests-py3.txt
+      install: pip install $(cat requirements-tests-py3.txt | grep flake8)
```
Comment (asottile) [inline .travis.yml]: UUOC: can just be `$(grep flake8 requirements-tests-py3.txt)`

**Comment by JelleZijlstra** [comment]
Should we land this soon? Currently not that many PRs are open, and I'd love to get this in so I don't have to worry about how to format stubs.

cc @CraftSpider and @hauntsaninja

**Comment by hauntsaninja** [comment]
Assuming we're doing this*, now seems like a good time to me too!

<details>

I know it's not the Black way, so hiding my small dissent. Despite liking it in code, I've personally come to dislike one line per arg in stubs. In particular, I find it much harder to compare long overloads (e.g. `stdlib/3/subprocess.pyi`). I can get over it, but thought I'd mention in case other people feel similarly.
</details>

**Comment by srittau** [comment]
@hauntsaninja Quick reply to your gripes: I am not totally on board with all formatting decisions of black (or PEP 8), either, but in the end it's more important to me to have consistent formatting than one that perfectly aligns with my tastes.

**Comment by CraftSpider** [comment]
It seems like we're not likely to get much lower, I'm down to land this soon.

<details>As for the formatting style, I also find one line per arg int stubs less readable, but consistency is generally more important. If most others also find it inconvenient, that makes it definitely a known papercut, but if that's the only one I think that's an acceptable trade-off for the consistency gained.</details>

**Comment by srittau** [comment]
We can land this once #4224 lands.

**Comment by JelleZijlstra** [comment]
We applied Black/isort in #4287. I'm going to fix up this PR and merge it after lunch (unless @hauntsaninja or @srittau gets to it first).

**Comment by JelleZijlstra** [comment]
Hm, isort is failing in CI, maybe it's not picking up our config from pyproject.toml?

**Comment by JelleZijlstra** [comment]
Ah ```/home/travis/virtualenv/python3.8.1/lib/python3.8/site-packages/isort/settings.py:296: UserWarning: Found /home/travis/build/python/typeshed/pyproject.toml with [tool.isort] section, but toml package is not installed. To configure isort with /home/travis/build/python/typeshed/pyproject.toml, install with 'isort[pyproject]'.```

**Comment by Eric-Arellano** [comment]
> but toml package is not installed. To configure isort with /home/travis/build/python/typeshed/pyproject.toml, install with 'isort[pyproject]'.

Rather than installing `toml` as a dedicated requirement, I believe the error message is saying to install `isort[pyproject]`, similar to how you can install `requests[security]`. Your requirements.txt would have `isort[pyproject]==4.3.21`.

**Comment by JelleZijlstra** [comment]
@Eric-Arellano you're right, I was too lazy to figure out how to combine the `[pyproject]` and `==version` syntax, but that's a better way. What I did should work too though because the pyproject extra just installs toml (https://github.com/timothycrosley/isort/blob/develop/pyproject.toml#L42).

**Comment by hauntsaninja** [comment]
Looks like isort is still failing

**Comment by JelleZijlstra** [comment]
OK, now there's another isort error, which is apparently because isort behaves differently depending on whether or not `typing-extensions` is installed in the current environment (?!?).

For now I'm going to remove isort from CI and land the Black part of his PR. We can try to figure out a less fragile isort config in a separate PR.

**Comment by JelleZijlstra** [comment]
CI passed, I'm landing this now. We can discuss how to turn on isort in #4288.

**Pull request merged**